devito/operator.py

```diff
@@ -94,6 +95,12 @@ def __init__(self, expressions, **kwargs):
         # Wrap expressions with Iterations according to dimensions
         nodes = self._schedule_expressions(clusters)
 
+        # Initialise Argument Engine
+        self.argument_engine = ArgumentEngine()
+
+        # Extract dimension offsets
+        self.argument_engine.extract_dimension_offsets([i.stencil for i in clusters])
```

Comment:
with this ``ArgumentEngine``, are you simply trying to extract the outermost offsets along each dimension? An example; given the expression
```
A[i-3, j+1] - B[i+4, j+5] + C[i-2, j+3]
```
Return:
```
i: (-3, 4)
j: (1, 5)
```

Reply:
Is there a question?

Reply:
yes a confirmation that this what you're trying to do

Reply:
Yes but it doesn't return those values, it caches them on the engine object.

Reply:
OK, then I think there should be ways of doing this without this engine, which is only over-complicating things

a few lines above you have the ``stencils`` variable, which you can inspect to get the max/min offsets along any Operator dimension (you might even add helper methods to ``Stencil`` directly?).  What happens if you try using this?

Reply:
any updates ?

Reply:
The ArgumentEngine will provide a runtime environment for the operator. I see only a cleaner interface because of its introduction - there is no overcomplication. The stencil radius is a property that is computed at compile (operator creation) time and used at runtime (apply). The ArgumentEngine is self-sufficient in what it needs to apply the operator and hence stores these properties on itself.

Reply:
> What happens if you try using this?

? -- point is, if this suffices, we can just store this info and forget about all the rest